src/main/scala/tsiHost/TsiHostWidget.scala

```diff
@@ -0,0 +1,171 @@
+package testchipip
+
+import chisel3._
+import chisel3.util._
+
+import freechips.rocketchip.config.{Parameters, Field}
+import freechips.rocketchip.subsystem.{BaseSubsystem}
+import freechips.rocketchip.regmapper.{HasRegMap, RegField}
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.util._
+
+import SerialAdapter._
+
+/**
+ * TSI Host parameter class
+ *
+ * @param mmioRegWidth size of the MMIO data being sent back and forth in bits
+ * @param serialIfWidth size of the serialIO *out* of the widget
+ * @param txQueueEntries size of the queue for sending TSI requests
+ * @param rxQueueEntries size of the queue for recieving TSI responses
+ * @param baseAddress start address of the MMIO registers
+ * @param tlAddressSet set of addresses that you can access on the target side
+ * @param tlRegionType is cache of TL transcations happening on this side?
+ * @param tlNumTransactions is the number of inflight transactions that can happen (also determines the id associated with that transaction)
+ * @param tlLineBytes you can transfer up to a line at a time (this determines the amount of beats in a TL burst)
+ * @param bypassMMIO disconnect the mmio and have a bypass port that connects directly to the SerialAdapter
+ */
+case class TSIHostParams(
+  mmioRegWidth: Int = 32,
+  serialIfWidth: Int = 32,
+  txQueueEntries: Int = 16,
+  rxQueueEntries: Int = 16,
+  baseAddress: BigInt = BigInt(0x10017000),
+  tlAddressSet: AddressSet = AddressSet(0, BigInt("FFFFFFFF", 16)),
+  tlRegionType: RegionType.T = RegionType.UNCACHED,
+  tlNumTransactions: Int = 1,
+  tlLineBytes: Int = 64,
+  bypassMMIO: Boolean = false
```

Comment:
Rather than use a bypass can you use a `BundleBridge` and a `TLPatternPusher` in your test to perform the TL transactions you need?

Reply:
Is there an example of `BundleBridge` used somewhere else? The goal of the bypass was to be able to still send TSI requests without going through the MMIO. The `SerialAdapter` would then get this and send/convert the TSI into TL. If I understand `TLPatternPusher` correctly, this is just used to create TL request sequence. That is already handled by the `SerialAdapter`.

Reply:
What I meant was that you could generate the transactions that you'd send to the MMIO. Otherwise I'm not a big fan of the boolean to add ios for simulation only. I'd say at that point just unit test the underlying (non-diplomatic) widgets and we'll have a larger integration test that makes sure the diplomacy stuff works. @colinschmidt ideas here?

Reply:
Offline Conversation: For now, create an inner block with a SerialIO interface to connect to (aka split the Frontend MMIO from the Backend Adapter + TLSerdesser). Unit test that block. Wait to work on adding a TLPatternPusher to the MMIO end.